MacrobenchmarkSample/app/src/main/java/com/example/macrobenchmark/target/util/SampleViewModel.kt

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright 2022 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.example.macrobenchmark.target.util
+
+import android.app.Application
+import android.util.Log
+import androidx.datastore.preferences.core.edit
+import androidx.datastore.preferences.core.stringPreferencesKey
+import androidx.lifecycle.AndroidViewModel
+import androidx.lifecycle.viewModelScope
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.map
+import kotlinx.coroutines.launch
+
+/** ViewModel for this sample */
+class SampleViewModel(application: Application) : AndroidViewModel(application) {
+
+    private val KEY_USER = stringPreferencesKey("user")
+    private val KEY_PASSWORD = stringPreferencesKey("password")
+
+    private val _data = SampleData()
+    private val _login = AppLogin()
+
+    val data: SampleData = _data
+    val login = _login
+
+    /**
+     * Attempts to log in with the provided credentials.
+     * @return `true` if successful, `false` otherwise.
+     */
+    fun login(userName: String, password: String) {
+        _login.userName = userName
+        _login.password = password
+
+        viewModelScope.launch {
+            val context = getApplication<Application>()
+            context.dataStore.edit { settings ->
+                settings[KEY_USER] = userName
+                settings[KEY_PASSWORD] = password
+                Log.d(TAG, "Wrote AppLogin to Data Store.")
+            }
```

Comment:
this feels like an implementation detail of the `AppLogin` class. What would you say to move it there?

Reply:
AppLogin holds the data but doesn't care where it's from. In a real app this should be part of the repository, so keeping it in the ViewModel looks good to me.